source/common/tcp_proxy/upstream.h

```diff
@@ -129,6 +157,9 @@ class HttpUpstream : public GenericUpstream, protected Http::StreamCallbacks {
     void decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override {
       if (!parent_.isValidResponse(*headers) || end_stream) {
         parent_.resetEncoder(Network::ConnectionEvent::LocalClose);
```

Comment:
I think we still have a weird corner case where we resetEncoder where if "inform_downstream" is true, we send an upstream event downstream before the pool knows it has an upstream associated.

I think we could simplify this by dropping onGenericPoolFailure below, and instead in resetEncoder, 
if(inform_downstream) {
if (deferrer) {
  deferrer.onGenericPoolFailure()
} else {
  former logic.
}
}
I think we always reset the encoder on failure (be it decodeHeaders failing, or disconnect) and then resetEncoder would take care of making it look like an event, or like a pool failure.  WDYT?  It's definitely worth an integration test and/or unit tests since it's tricky timing.

Reply:
yep,  I saw your previous comment and I was still figuring out how to do it. I pushed what I did so far, not tested yet. If you have the time to take a look let me know if I'm on the good path ;-)